openimageio: 2.5.16.0 -> 3.0.5.0

yzx9 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Openimageio v3 was released in Nov 9, 2024. This PR bumps openimageio from `2.5.16.0` to `3.0.5.0`.

However, v3 introduces many breaking changes that some consumers are not yet ready for, and updating them would require complex changes. To address this, we’ve introduced `openimageio_2` at version `2.5.17.0`. This legacy package is expected to be removed once it is no longer used in nixpkgs.

Also, openimageio has been [relicensed under Apache-2.0](https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/3905)

close #363100

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

yzx9 committed:
openimageio: move to by-name

amarshall reviewed (commented):
Thanks! (Comments based on diffs only, have not yet done build testing)

I have a mild preference for the commits to be something like:

- openimageio: move to by-name
- openimageio: 2.5.16.0 -> 2.5.17.0, fix incorrect license
- openimageio: 2.5.17.0 -> 3.0.5.0; openimageio_2: init at 2.5.17.0

With all the necessary changes to dependents in that last one. This is to make bisecting easier—with the current commits in the PR the majority of commits have various drvs (presumably) failing to build.

yzx9 committed:
openimageio: 2.5.16.0 -> 2.5.17.0, fix incorrect license

- v2.5.17.0 was released on Nov 1, 2024, includes bug fixes
- openimageio relicense their code under Apache 2.0
  https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/3905
- remove unnecessary patch
- remove `rec` and `with lib;`

yzx9 committed:
openimageio: 2.5.17.0 -> 3.0.5.0; openimageio_2: init at 2.5.17.0

Openimageio v3 was released in Nov 9, 2024. However, v3 introduces many
breaking changes that some consumers are not yet ready for, and
updating them would require complex changes. To address this, we’ve
introduced openimageio_2 at version 2.5.17.0. This legacy package is
expected to be removed once it is no longer used in nixpkgs.

yzx9 committed:
embree{,2}: use openimageio_2

yzx9 committed:
osl: use openimageio_2

yzx9 committed:
gpupad: use openimageio_2

yzx9 committed:
python3Packages.materialx: use openimageio_2

yzx9 committed:
python3Packages.openusd: use openimageio_2

yzx9 commented on pkgs/by-name/op/openimageio/package.nix:
```diff
@@ -15,21 +15,21 @@
   postPatch = ''
     substituteInPlace ./src/cmake/Config.cmake.in \
-      --replace " @ZLIB_VERSION@" ""
+      --replace-warn " @ZLIB_VERSION@" ""
```
Apologies for the incomplete check. It turns out this patch is no longer needed, so I’ve removed it.

yzx9 commented on pkgs/by-name/op/openimageio/2.nix:
```diff
@@ -0,0 +1,15 @@
+
+openimageio.overrideAttrs (
+  finalAttrs: prevAttrs: {
+    name = "${prevAttrs.pname}-${finalAttrs.version}";
```
Got it!

yzx9 commented on pkgs/by-name/op/openimageio/package.nix:
```diff
@@ -76,8 +76,8 @@ stdenv.mkDerivation rec {
     description = "Library and tools for reading and writing images";
-    license = licenses.bsd3;
-    maintainers = [ ];
+    license = licenses.asl20;
```
Got it!

yzx9 committed:
blender: use openimageio_2

yzx9 commented:
@amarshall Thanks for your review! bisecting is reasonable to make each commit build success. I've update the commits. Also, there are still some build failure I am try to addressed. And the [gpupad look like has been broken in a while](https://hydra.nixos.org/search?query=gpupad)

SuperSandro2000 reviewed (approved):
LGTM

gador commented:
The tag seems to have changed
```
source> trying https://github.com/AcademySoftwareFoundation/OpenImageIO/archive/refs/tags/v3.0.5.0.tar.gz
source>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
source>                                  Dload  Upload   Total   Spent    Left  Speed
source>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
source> 100 50.2M    0 50.2M    0     0  5084k      0 --:--:--  0:00:10 --:--:-- 5319k
source> unpacking source archive /private/tmp/nix-build-source.drv-0/download.tar.gz
error: hash mismatch in fixed-output derivation '/nix/store/r0mv9p5ldh5g121qpnq6chlqz8n0lvav-source.drv':
         specified: sha256-d5LqRcqWj6E9jJYY/Pa5e7/MeuQGMjUo/hMCYRKsKeU=
            got:    sha256-CG2tb6jf7xmQnrLaTqEFKG3g4nkEwZoI44bXxdCRmrM=
```

yzx9 commented:
@gador Thanks for your review! The tag itself wasn’t changed. However, a subtle mistake was made when I changed
`openimageio.overrideAttrs(finalAttrs: prevAttrs: { ... })` to `openimageio.overrideAttrs (finalAttrs: { ... })`. In this case, `finalAttrs` is actually bound to `prevAttrs`, which caused us to reference the wrong version. To avoid this, I’ve switched to using the `let ... in` pattern.

gador commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 401089`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>blender</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 15 packages built:</summary>
  <ul>
    <li>embree</li>
    <li>materialx (python312Packages.materialx)</li>
    <li>openimageio</li>
    <li>openimageio.bin</li>
    <li>openimageio.dev</li>
    <li>openimageio.doc</li>
    <li>openimageio_2</li>
    <li>openimageio_2.bin</li>
    <li>openimageio_2.dev</li>
    <li>openimageio_2.doc</li>
    <li>openusd</li>
    <li>osl</li>
    <li>python312Packages.openusd</li>
    <li>python313Packages.materialx</li>
    <li>python313Packages.openusd</li>
  </ul>
</details>
